locator/abstract_replication_strategy.cc

```diff
@@ -178,6 +178,9 @@ abstract_replication_strategy::get_ranges(inet_address ep, token_metadata& tm) c
     for (auto tok : tm.sorted_tokens()) {
         for (inet_address a : calculate_natural_endpoints(tok, tm)) {
             if (a == ep) {
+                if (seastar::thread::running_in_thread()) {
+                    seastar::thread::maybe_yield();
+                }
                 insert_token_range_to_sorted_container_while_unwrapping(prev_tok, tok, ret);
```

Comment:
This is dangerous. A function can happen to be in a thread by accident.

```c++
    return std::async([] {
         // in a thread now
         func_that_returns_a_future().then([] {
              // not in a thread, bit running_in_thread() returns true
              int x = 3; // starts life in a thread stack
              seastar::thread::maybe_yield();
              // thread stack may be gone by now, or we may execute in a different thread
              print(x); // maybe not 3
         }).forward_to(some_promise);
    });

Reply:
It is better to return a future. And even better to fix the function not to take 16 seconds. What is it computing, all the digits of pi?

Reply:
Good example. Can we somehow tell if the code is really under a thread?

Reply:
abstract_replication_strategy::get_ranges() loops all the vnode tokens, calculates natural_endpoints for each of the tokens to check if the node owns the range.

Reply:
> Good example. Can we somehow tell if the code is really under a thread?

Only by code inspection. At runtime Scylla kills this information.

Reply:
> abstract_replication_strategy::get_ranges() loops all the vnode tokens, calculates natural_endpoints for each of the tokens to check if the node owns the range.

Why does it take 16 seconds?

Reply:
Note, the 16 seconds is not contributed only by the get_ranges. Also the copy_if: 

```
739         std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(sstables), [&owned_ranges, schema] (const sstables::shared_sstable& sst) {
740             return owned_ranges.empty() || needs_cleanup(sst, owned_ranges, schema);
741         });
```

Reply:
The copy_if itself is O(n) with a moderate N. I expect the problem is in conjunction with needs_cleanup().

Reply:
Yes,  it iterates the whole owned ranges. 

```
711 static bool needs_cleanup(const sstables::shared_sstable& sst,
712                    const dht::token_range_vector& owned_ranges,
713                    schema_ptr s) {
714     auto first = sst->get_first_partition_key();
715     auto last = sst->get_last_partition_key();
716     auto first_token = dht::get_token(*s, first);
717     auto last_token = dht::get_token(*s, last);
718     dht::token_range sst_token_range = dht::token_range::make(first_token, last_token);
719 
720     // return true iff sst partition range isn't fully contained in any of the owned ranges.
721     for (auto& r : owned_ranges) {
722         if (r.contains(sst_token_range, dht::token_comparator())) {
723             return false;
724         }
725     }   
726     return true;
727 }
```